Mod View cursors, modHistory

Henry-Graves opened the pull request:
# Describe your changes
Add modHistory database entries on all relevant moderation actions.
Add types & IQs for reports & modHistory.

# Issue ticket number and link
closes all main reporting functionality under le epic #105

JakeSteinebronn commented:
I think we should consider doing some simple grouping based on the day, for mod actions. Cuz like, there could be a lot of actions happening in a day, and this collection could get quite chunky. So like instead, make the documents called like 
day_month_year and then it has a list called `reports` to which you can `arrayUnion` with a new string or object or whatever. What do you think?

Henry-Graves commented:
Yeah that could be good especially if an employee wants to scroll back to a mod action that happened months ago, we could display a dropdown of each day action was taken. They'd be able to scroll down faster and we wouldn't have to do as many reads to get there

JakeSteinebronn commented:
Status on this PR?

Henry-Graves commented:
@JakeSteinebronn it's in progress, having some trouble with this part though:

The collection modHistory is gonna have docs in the format MM_DD_YYYY and those docs have a list of modActions taken on that day. I have all the logic around making a doc for the day and updating modActions, but I'm struggling with making an IQ.

I'm trying to set up the IQ so that it only fetches the dates actions took place (to save on reads), then from there employees can click on a date to get the rest of the data for that doc (the list of actions taken that day).

It seems like the right way to set it up is to make the modActions list a sub-collection of each doc. An IQ on modHistory won't read any sub-collections, then we could have an IQ or a .getAll() for the modActions.

Henry-Graves commented:
the last commit does the following: 
- lots of linting, sorry it's hard to read
- fix IQ keys
- rename type ModHistory -> ModAction
- refactor ModHistory entries using User.addModAction()

db looks like this
![image](https://user-images.githubusercontent.com/73561858/219492842-632fcc0d-7eb1-4f69-874e-2429bf7b103d.png)

JakeSteinebronn commented on queries/modHistory.ts:
```diff
@@ -0,0 +1,33 @@
+
+export function getIQParams_ModHistory() {
+  return {
+    queryKey: ['modHistory', 'AAAAAAAAAAAAAAAAHHHHHHHHHHHH'], // 2nd param doesn't matter here bc no parent collection for modHistory
```
You dont need to have the aahhhh; the queryKey can be an array of any number of strings, arrays, or objects and thus simply `['modHistory']` suffices

JakeSteinebronn commented on queries/report.ts:
```diff
@@ -0,0 +1,33 @@
+
+export function getIQParams_Reports() {
+  return {
+    queryKey: ['reports', 'AAAAAAAAAAAAAAAAHHHHHHHHHHHH'], // 2nd param doesn't matter here bc no parent collection for reports
```
same here, no ahhhhhh needed

JakeSteinebronn commented on types/modAction.ts:
```diff
@@ -0,0 +1,71 @@
+
+export class ModAction extends LazyObject {
+  public userModerated?: User;
+  public mod?: User;
```
I'd appreciate the comments in the same style as the other lazy objects, the ones saying which ones are required and which ones are optional. 

Also, could we rename `mod` to `moderator`? It seems like `mod` can be a noun or a verb, and `mod` interpreted as a verb might mean the thing reported, not the person doing the moderating

One last thing, maybe I'm missing something - why don't we have a ref to the moderated content here?

JakeSteinebronn replied:
oh I see, i'm stupid, if the content is deleted then what would the reference be to. :p

JakeSteinebronn commented on types/report.ts:
```diff
@@ -0,0 +1,68 @@
+  public initWithDocumentData(data: DocumentData): void {
+    this.reporter = new User(data.user);
+    this.reported = new User(data.user);
+    if (data.content instanceof Post) this.content = new Post(data.mod);
```
Did you test this? no way this works... right? `data.content` is a documentReference. If you tested this and it works, then disregard, otherwise I think this can be handled in one of two ways:

- just have 3 separate optional fields repotedPost, reportedUser, reportedComment
- data.content is a documentReference, so you can get the path in that documentReference and check if the collection is `users/`, `comments/`, or `posts/`

I'm generally reviewing these PRs assuming you've tested them, so i'm hoping that's the case, and this works as intended, but if so, i'm extremely surprised

Henry-Graves replied:
this might be hard to read but this is some peace of mind for you showing i did test. I usually do, i just forgor on the initWithDocumentDatas the other day
![Screenshot (149)](https://user-images.githubusercontent.com/73561858/219894090-0c998d2f-8754-4069-a326-fb25d893915b.png)

JakeSteinebronn replied:
no prob man <3 yeah that's fuckin impossible to read :p

JakeSteinebronn commented on types/user.ts:
```diff
@@ -323,19 +335,66 @@ export class User extends LazyObject {
+  public async addModAction(userModerated: User, mod: User, modReason: String) {
+    const documentId = await this.getDateMM_DD_YYYY();
+    const modHistoryDocRef = await doc(db, 'modHistory', documentId);
+    const modHistorySnap = await getDoc(modHistoryDocRef);
```
I can clean this up if you don't want to, but there's a way to update a document if it exists else create it, with `setDoc(docRef, content, {merge: true})`

JakeSteinebronn reviewed (changes requested):
Requesting changes since I don't understand how the instanceof thing can work - if you've tested it and it correctly identifies a documentReference as instanceof T extends lazyobject, could you please explain how that works to me? Thanks!

Henry-Graves commented:
Big commit there, here's everything done:
- updated modAction to account for cases where the users in the action log have been deleted. ModAction stores a ref to the user, if they get deleted we should still have some way of knowing who that user was. We used to store their email for this, but now we're storing their username. With this, web would check if the user no longer exists, then fill in the UI with their stored username if necessary.
- removed unnecessary 2nd param on IQ queryKeys
- added comments to type fields describing data presence
- changed all instances of 'mod' to 'moderator'
- changed User.addModAction() to use setDoc {merge: true}
- fixed initWithDocumentData() for Report & ModAction

JakeSteinebronn commented on types/modAction.ts:
```diff
@@ -1,40 +1,89 @@
       data.timestamp.seconds * 1000 + data.timestamp.nanoseconds / 1000000
     );
 
+    // check if these users still exist in db
```
clever :) I forgot about this kinda thing
